src/hotspot/share/prims/jvmtiTagMapTable.hpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_VM_PRIMS_TAGMAPTABLE_HPP
+#define SHARE_VM_PRIMS_TAGMAPTABLE_HPP
+
+#include "oops/weakHandle.hpp"
+#include "utilities/hashtable.hpp"
+
+class JvmtiEnv;
+
+// Hashtable to record oops used for JvmtiTagMap
+class JvmtiTagMapEntryClosure;
+
+class JvmtiTagMapEntry : public HashtableEntry<WeakHandle, mtServiceability> {
```

Comment:
So looking at the concurrent hashtable, the entries can be created without saving the hashcode.  I was going to use that at first but didn't want to cut/paste the boilerplate to do so and the jvmti tag map hashtable is always accessed with a lock.  This could be a future RFE if necessary and would also serve to eliminate another ad-hoc hashtable.